include/pybind11/pybind11.h

```diff
@@ -223,20 +231,48 @@ class cpp_function : public function {
         }
     }
 
+    // Utility class that keeps track of all duplicated strings, and cleans them up in its destructor,
+    // unless they are released. Basically a RAII-solution to deal with exceptions along the way.
+    class strdup_guard {
+    public:
+        ~strdup_guard() {
+            for (auto s : strings)
+                std::free(s);
+        }
+        char *operator()(const char *s) {
+            auto t = strdup(s);
+            strings.emplace_back(t);
```

Comment:
Any reason to use `emplace_back` instead of `push_back` here?

Reply:
Nope (except for maybe `clang-tidy`'s `modernize-use-emplace`?) Any reason for `push_back`?

Reply:
It's a simpler mechanism when you already have a `T` and it also doesn't need to instantiate a variadic member function template. On the other hand `emplace()` can't provide its main advantage of constructing in place, because you're already passing a `T`.

Not a big deal either way. Some people prefer to use `push_back(t)` while others prefer the simpler rule of "always use `emplace`".

Reply:
Yeah, OK. I do agree, actually, if it's a pointer, but something tells me clang-tidy bit me, last time I tried it...

@henryiii, any preferences? You're the one that set up clang-tidy :-P

Reply:
I think clang-tidy would be fine with `push_back()` here, though I'm not really against leaving it as-is.

Reply:
If clang-tidy wants emplace_back, then that's fine with me, though I think it should be happy with push_back for a pointer like this.

Reply:
Let me try. I don't think I actually tried this out.

Reply:
OK, you're right; clang-tidy passes! I must have misremembered/misinterpreted an old clang-tidy failure. This is better, indeed :-)